frontend

aryanbhosale opened the pull request:
# Pull Request

## Description

A clean dashboard for the user to view their solar generation forecasts


https://github.com/user-attachments/assets/cd026503-a1f3-4735-b292-698f179ca79e



## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

Sukh-P commented on frontend/app.py:
```diff
@@ -0,0 +1,256 @@
+
+st.title("☀️ Open Source Quartz Solar Forecast")
+
+def get_enphase_auth_url():
```
I may be missing something so let me know if I am, is there a reason why we are duplicating a bunch of these functions here and not importing them from the `enphase.py` file?

aryanbhosale replied:
> I may be missing something so let me know if I am, is there a reason why we are duplicating a bunch of these functions here and not importing them from the `enphase.py` file?

This is because some of these functions in enphase.py and even the forecast.py except inputs in the CLI, so reusing them doesn't store the custom inputs, here for example the auth URL, in the browser. So to remove the CLI part entirely to make it browser based I modified those functions a bit

aryanbhosale replied:
> > I may be missing something so let me know if I am, is there a reason why we are duplicating a bunch of these functions here and not importing them from the `enphase.py` file?
> 
> This is because some of the functions in enphase.py and even in forecast.py accept inputs in the CLI, so reusing them doesn't store the custom inputs, here for example the auth URL, in the browser. So to remove the CLI part entirely to make it browser based I modified those functions a bit and rewrote them in the new streamlit file

Sukh-P commented on frontend/app.py:
```diff
@@ -0,0 +1,256 @@
+            timestamp_str = datetime.fromtimestamp(timestamp, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
+            ts = pd.to_datetime(timestamp_str)
+
+            site_live = PVSite(latitude=51.75, longitude=-1.25, capacity_kwp=1.25, inverter_type="enphase")
```
I think as Peter mentioned it would be very nice if these (lat, long and capacity) could be user inputs via the UI! Although to be fair that could be in the next PR

aryanbhosale replied:
> I think as Peter mentioned it would be very nice if these (lat, long and capacity) could be user inputs via the UI! Although to be fair that could be in the next PR

Since it's a small feature I could add it in here too

Sukh-P replied:
@aryanbhosale are you still planning on doing this in this PR?

aryanbhosale replied:
> @aryanbhosale are you still planning on doing this in this PR?

another PR would do I think, just so that we dont overload this one. Since the CLI version also doesnt use custom values and this PR was just to provide a UI to visualize what the CLI code did. So another PR for the custom values would make sense because it's not just limited to user inputs, we'll need to put sanity checks too

Sukh-P commented:
@aryanbhosale in terms of folder structure I was thinking that this was perhaps better to have this in the example folder? But keen to hear your thoughts, just that way I think it is closer to the inverter example code

aryanbhosale commented:
> @aryanbhosale in terms of folder structure I was thinking that this was perhaps better to have this in the example folder? But keen to hear your thoughts, just that way I think it is closer to the inverter example code

Yes true, but I only kept it in the root directory because usually we separate the frontend and backend in 2 folders, and the different functions in the quartz_solar_forecast directory(spread across a several files and folders) act as the backend to this, so I thought that it should be at the same level to the  quartz_solar_forecast directory

aryanbhosale commented:
> > @aryanbhosale in terms of folder structure I was thinking that this was perhaps better to have this in the example folder? But keen to hear your thoughts, just that way I think it is closer to the inverter example code
> 
> Yes true, but I only kept it in the root directory because usually we separate the frontend and backend in 2 folders, and the different functions in the quartz_solar_forecast directory(spread across several files and folders) act as the backend to this, so I thought that it should be at the same level as the quartz_solar_forecast directory

aryanbhosale commented:
Hey @Sukh-P , I've improved the error handling and the auth workflow in my latest commit. The issue was that before this the app was requesting a new authorization code for each forecast run, leading to "Invalid authorization code" errors and there was no persistent storage of the access token between interactions.

So to make it work, I implemented Streamlit's session state to store the access token, allowing it to persist across multiple forecast runs and modified the authorization flow to check for an existing access token before requesting a new one.

peterdudfield commented on frontend/app.py:
```diff
@@ -0,0 +1,275 @@
+    da = process_pv_data(live_generation_kw, ts, site)
+    return da
+
+def predict_ocf(site: PVSite, model=None, ts: datetime | str = None, nwp_source: str = "icon", access_token: str = None, enphase_system_id: str = None):
```
Could you just import this function? 

https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/forecast.py#L10
you might have to add enphase tokens to this function

aryanbhosale replied:
> Could you just import this function?
> 
> https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/forecast.py#L10 you might have to add enphase tokens to this function

that function indirectly uses the make_pv_data from data.py, that uses some of the CLI functions from enphase.py, since we want all the user inputs to be only on the web, i have modified these functions in this file

peterdudfield commented on frontend/app.py:
```diff
@@ -0,0 +1,275 @@
+    pred_df = forecast_v1_tilt_orientation(nwp_source, nwp_xr, pv_xr, ts, model=model)
+    return pred_df
+
+def predict_tryolabs(site: PVSite, ts: datetime | str = None):
```
same as above, couldnt you import this function?

aryanbhosale replied:
> same as above, couldnt you import this function?

yes my bad, didnt realise in the flow that this doesnt use the CLI functions

aryanbhosale replied:
> same as above, couldnt you import this function?

done in my latest commit

Sukh-P commented on frontend/app.py:
```diff
@@ -0,0 +1,247 @@
+    pred_df = forecast_v1_tilt_orientation(nwp_source, nwp_xr, pv_xr, ts, model=model)
+    return pred_df
+
+def run_forecast(site: PVSite, model: str = "gb", ts: datetime | str = None, nwp_source: str = "icon", access_token: str = None, enphase_system_id: str = None) -> pd.DataFrame:
```
This function can be imported from the `forecast.py` file instead of duplicated from here?

aryanbhosale replied:
> This function can be imported from the `forecast.py` file instead of duplicated from here?

This function uses the predict_ocf function that uses make_pv_data from data.py, which uses some CLI based functions from enphase.py, so to avoid that I modified this here

Sukh-P commented:
> > @aryanbhosale in terms of folder structure I was thinking that this was perhaps better to have this in the example folder? But keen to hear your thoughts, just that way I think it is closer to the inverter example code
> 
> Yes true, but I only kept it in the root directory because usually we separate the frontend and backend in 2 folders, and the different functions in the quartz_solar_forecast directory(spread across a several files and folders) act as the backend to this, so I thought that it should be at the same level to the quartz_solar_forecast directory

So I agree that makes sense in terms of the split, perhaps we should rename the folder though to steamlit_app or something like that so it's a bit more clear what it is

aryanbhosale commented:
> > > @aryanbhosale in terms of folder structure I was thinking that this was perhaps better to have this in the example folder? But keen to hear your thoughts, just that way I think it is closer to the inverter example code
> > 
> > 
> > Yes true, but I only kept it in the root directory because usually we separate the frontend and backend in 2 folders, and the different functions in the quartz_solar_forecast directory(spread across a several files and folders) act as the backend to this, so I thought that it should be at the same level to the quartz_solar_forecast directory
> 
> So I agree that makes sense in terms of the split, perhaps we should rename the folder though to steamlit_app or something like that so it's a bit more clear what it is

That makes sense, but what you'd mentioned earlier sounds better, that it'll stay close to the parent function that it helps visualize, ie inside the examples folder. Should I shift it there without changing the name?

peterdudfield commented on examples/streamlit_app/app.py:
```diff
@@ -0,0 +1,245 @@
+            site_live = PVSite(latitude=51.75, longitude=-1.25, capacity_kwp=1.25, inverter_type="enphase")
+            site_no_live = PVSite(latitude=51.75, longitude=-1.25, capacity_kwp=1.25)
+
+            predictions_with_recent_pv_df = run_forecast(site=site_live, ts=ts, access_token=access_token, enphase_system_id=enphase_system_id)
```
has run foreacst been updated to take `access_token` and `enphase_system_id`

aryanbhosale replied:
> has run foreacst been updated to take `access_token` and `enphase_system_id`

ive modified the run_forecast function in the app.py of streamlit, so that it doesnt use the CLI based functions

peterdudfield commented:
Could you add screen shot of the UI in the PR please?

aryanbhosale commented on examples/streamlit_app/app.py:
```diff
@@ -0,0 +1,245 @@
+            timestamp_str = datetime.fromtimestamp(timestamp, tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S')
+            ts = pd.to_datetime(timestamp_str)
+
+            site_live = PVSite(latitude=51.75, longitude=-1.25, capacity_kwp=1.25, inverter_type="enphase")
```
> Could you put a Todo in the code here, about adding these values as inputs into the UI. For example `latitude` `longitude` and capacity_kw

Okay sure

aryanbhosale commented:
> Could you add screen shot of the UI in the PR please?

ive added a video

Sukh-P reviewed (approved):
Nice work adding this streamlit frontend in!
